### PR TITLE
Irix support (libultra_rom/libultra OK)

### PR DESCRIPTION
**Pull request by hensldm**
Adds support for libultra* builds (though libultra_d is still excluded).

Main design decision that should be looked at is splitting the Makefile into three. One main one for the actually building with all the commands, and than one each for GCC and IDO defines. I liked it better than the Swiss cheese if else's I had at first, but definitely would like others opinions.

Other decisions I'll comment directly in the PR.

Note: This does not match .mdebug, so copys all the object files to a .mdebug in the base folder, than strips the .mdebug section from the extracted object files. This also means the archive can't be matched directly. I still used Tharo's `ar.py` tools becuase why not, though had to adjust a few things since some files were empty except for the .mdebug, so after stripping it the symtab also got stripped as it had nothing left.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -13,24 +15,19 @@ WORKING_DIR := $(shell pwd)
...
-GBIDEFINE := -DF3DEX_GBI_2
-CPPFLAGS = -D_MIPS_SZLONG=32 -D__USE_ISOC99 $(GBIDEFINE)
-INCLUDES = -I . -I $(WORKING_DIR)/include -I $(WORKING_DIR)/include/gcc -I $(WORKING_DIR)/include/PR
+ifeq ($(findstring libgultra,$(TARGET)),libgultra)
+-include Makefile.gcc
+else ifeq ($(findstring libultra,$(TARGET)),libultra)
+-include Makefile.ido
+else
+$(error Invalid Target)
+endif
```
Comment (hensldm) [inline Makefile:25]: As mentioned this is the main decision choice to split the makefile.

File: Makefile [hunk 2]
```diff
@@ -59,7 +60,7 @@ endif
 
 BASE_OBJS := $(wildcard $(BASE_DIR)/*.o)
 # Try to find a file corresponding to an archive file in any of src/ asm/ or the base directory, prioritizing src then asm then the original file
-AR_ORDER = $(foreach f,$(shell $(AR) t $(BASE_AR)),$(shell find $(BUILD_DIR)/src $(BUILD_DIR)/asm $(BUILD_DIR)/$(BASE_DIR) -name $f -type f -print -quit))
+AR_ORDER = $(foreach f,$(shell $(AR) t $(BASE_AR)),$(shell find $(BUILD_DIR)/src $(BUILD_DIR)/asm $(BUILD_DIR)/$(BASE_DIR) -iname $f -type f -print -quit))
```
Comment (hensldm) [inline Makefile:63]: Note I had to change to `-iname` since some of the object files in irix builds are mixed case. Since we don't match the archive completely anyway, this seemed like the best solution rather than renaming the files, and dealing with lower-casing for gcc.

File: src/debug/assert.c [hunk 3]
```diff
@@ -1,5 +1,10 @@
-#include "PR/os_internal.h"
+#include "os.h"
+#include "../os/osint.h"
+#include "osint_debug.h"
+
+void __assertBreak(void);
 
 void __assert(const char* exp, const char* filename, int line) {
     osSyncPrintf("\nASSERTION FAULT: %s, %d: \"%s\"\n", filename, line, exp);
+    __assertBreak;
```
Comment (hensldm) [inline src/debug/assert.c:9]: I know this is back, but it actually appears in `libultra_d`'s symtab so must exist. But yeah it doesn't do anything AFAIK.

File: src/libc/bcmp.s [hunk 4]
```diff
@@ -3,8 +3,12 @@
 #include "sys/regdef.h"
 
 .text
-WEAK(_bcmp, bcmp)
+#ifdef __sgi
+WEAK(bcmp, _bcmp)
+LEAF(_bcmp)
+#else
 LEAF(bcmp)
```
Comment (hensldm) [inline src/libc/bcmp.s:10]: Not 100% sure this is the best way to handle this, though seems the most clear.
Basically GCC has `bcmp` as the strong symbol, but IRIX has `_bcmp` as the strong symbol (i.e. the actual function) with `bcmp` as the weak alias.
Note this applies to all three libc asm files.

File: src/os/getfpccsr.s [hunk 5]
```diff
@@ -7,4 +7,7 @@ LEAF(__osGetFpcCsr)
     STAY2(cfc1 v0, fcr31)
     jr ra
 END(__osGetFpcCsr)
+
+#ifndef __sgi
```
Comment (hensldm) [inline src/os/getfpccsr.s:11]: Again not 100% sure this is the best way, but GCC exports __osGetSR, while IDO does not.

File: src/os/maptlb.s [hunk 6]
```diff
@@ -9,7 +10,53 @@
 #define oddpaddr 16(sp)
 #define asid 20(sp)
 .text
+.set noreorder
 LEAF(osMapTLB)
+#if defined _DEBUG && defined __sgi
```
Comment (hensldm) [inline src/os/maptlb.s:15]: Just want to note that this piece seems to only appear in `libultra_d`.

File: tools/libelf.py [hunk 7]
```diff
@@ -1102,9 +1102,9 @@ def init_section(i):
             if s.sh_type == SHT_SYMTAB:
                 assert not symtab , "Found more than one symtab section?"
                 symtab = s
-        assert symtab is not None , "Could not find symtab"
         self.symtab = symtab
-        self.symtab.late_init()
+        if self.symtab is not None:
```
Comment (hensldm) [inline tools/libelf.py:1106]: No longer assumes the symtab exists (stripping off empty files that only have .mdebug files, also strips the symtab since it is now empty). Does mean I had to add a few not none checks here and in ar.py.

File: Makefile [hunk 8]
```diff
@@ -51,6 +48,10 @@ MARKER_FILES := $(O_FILES:.o=.marker)
 ifneq ($(NON_MATCHING),1)
 COMPARE_OBJ = cmp $(BASE_DIR)/$(@F:.marker=.o) $(@:.marker=.o) && echo "$(@:.marker=.o): OK"
 COMPARE_AR = cmp $(BASE_AR) $@ && echo "$@: OK"
+ifeq ($(COMPILER),ido)
+# Since we don't match .mdebug, we can't match the archive
+COMPARE_AR = echo "$@: OK"
+endif
```
Comment (AngheloAlf) [inline Makefile:54]: This means this will always print an OK even if the `.text`, `.data`, `.rodata` or `.bss` sections don't match?

File: src/io/aisetfreq.c [hunk 9]
```diff
@@ -1,6 +1,8 @@
 #include "PR/rcp.h"
 #include "PR/ultraerror.h"
 #include "../os/osint.h"
+//not sure if this should be here
+extern s32 osViClock;
```
Comment (AngheloAlf) [inline src/io/aisetfreq.c:5]:
```suggestion
// TODO: not sure if this should be here
extern s32 osViClock;
```

This way is easier to find stuff like this

File: src/os/parameters.s [hunk 10]
```diff
@@ -12,5 +12,6 @@ ABS(osCicId, 0x80000310)
 ABS(osVersion, 0x80000314)
 ABS(osMemSize, 0x80000318)
 ABS(osAppNMIBuffer, 0x8000031c)
-.fill 0x60
+
+.space 0x60
```
Comment (AngheloAlf) [inline src/os/parameters.s:16]: Why the change from `.fill` to `.space`?

File: src/os/setsr.s [hunk 11]
```diff
@@ -5,6 +5,6 @@
 .text
 LEAF(__osSetSR)
     STAY2(mtc0 a0, C0_SR)
-    nop
+    NOP
```
Comment (AngheloAlf) [inline src/os/setsr.s:8]: NOP

File: src/debug/assert.c (hunk 3 above)
Comment (AngheloAlf) [reply src/debug/assert.c:9]: I wonder if we should add a note here explaining what's going on

File: Makefile (hunk 8 above)
Comment (hensldm) [reply Makefile:54]: This is only when comparing the archives directly together. OBJ_COMPARE is still the same and will check that the object files built are the same.

File: src/os/initialize_isv.c [hunk 12]
```diff
@@ -113,15 +111,17 @@ int __checkHardware_isv(void) {
     gISVChk = 0;
 
     for (i = 0; i < 4; i++) {
-        osEPiReadIo(hnd, 0xB0000100 + i * 4, save + i);
+        osEPiReadIo(hnd, 0xB0000100 + i * 4, &save[i]);
     }
 
-    // data = 0;
+#ifndef __GNU__ // BUG: data is used uninitialized for GCC
```
Comment (hensldm) [inline src/os/initialize_isv.c:117]: It had been discussed before, but I really don't know how to reconcile this difference. Ido explicitly sets data to 0, but GCC goes crazy if you do.

File: Makefile.ido [hunk 13]
```diff
@@ -37,8 +36,8 @@ $(BUILD_DIR)/src/mgu/%.marker: ASOPTFLAGS := -O2
 endif
 
 $(BUILD_DIR)/src/os/initialize_isv.marker: OPTFLAGS := -O2
-$(BUILD_DIR)/src/log/delay.marker: MIPS_VERSION := -mips1
-$(BUILD_DIR)/src/libc/ll.marker: MIPS_VERSION := -mips3 -32
-$(BUILD_DIR)/src/libc/llbit.marker: MIPS_VERSION := -mips3 -32
-$(BUILD_DIR)/src/libc/llcvt.marker: MIPS_VERSION := -mips3 -32
-$(BUILD_DIR)/src/os/exceptasm.marker: MIPS_VERSION := -mips3 -32
+$(BUILD_DIR)/src/log/delay.marker: MIPS_VERSION := -mips1 -o32 -KPIC
```
Comment (AngheloAlf) [inline Makefile.ido:39]: Oh wonderful, a `-KPIC` file

File: src/os/getfpccsr.s [hunk 14]
```diff
@@ -6,8 +6,4 @@
 LEAF(__osGetFpcCsr)
     STAY2(cfc1 v0, fcr31)
     jr ra
-END(__osGetFpcCsr)
-
-#ifndef __sgi
-.globl __osGetSR
-#endif
+END(__osGetSR)
```
Comment (AngheloAlf) [inline src/os/getfpccsr.s:9]: Should we add a `//! bug` note?

File: src/os/setfpccsr.s [hunk 15]
```diff
@@ -7,8 +7,4 @@ LEAF(__osSetFpcCsr)
     STAY2(cfc1 v0, fcr31)
     STAY2(ctc1 a0, fcr31)
     jr ra
-END(__osSetFpcCsr)
-
-#ifndef __sgi
-.globl __osSetSR
-#endif
+END(__osSetSR)
```
Comment (AngheloAlf) [inline src/os/setfpccsr.s:10]: Same here

File: Makefile.ido [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+PATCH_AR_FLAGS := 40001 110 100644
+STRIP =
+
+export COMPILER_PATH := $(WORKING_DIR)/tools/ido
+
+CFLAGS := -c -Wab,-r4300_mul -G 0 -nostdinc -Xcpluscomm -fullwarn -woff 516,649,838,712
+ASFLAGS := -c -Wab,-r4300_mul -G 0 -nostdinc -woff 516,649,838,712
+CPPFLAGS = -D_MIPS_SZLONG=32 $(GBIDEFINE)
+IINC = -I $(WORKING_DIR)/include -I $(WORKING_DIR)/include/ido -I $(WORKING_DIR)/include/PR
+MIPS_VERSION := -mips2 -o32 -non_shared
```
Comment (Thar0) [inline Makefile.ido:14]: `-non_shared` doesn't really belong here

File: Makefile.ido [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+$(BUILD_DIR)/src/libc/llbit.marker: OPTFLAGS := -O1
+$(BUILD_DIR)/src/libc/llcvt.marker: OPTFLAGS := -O1
+$(BUILD_DIR)/src/log/%.marker: OPTFLAGS := -O1
+
+$(BUILD_DIR)/src/libc/%.marker: ASOPTFLAGS := -O2
+$(BUILD_DIR)/src/mgu/%.marker: ASOPTFLAGS := -O2
+endif
+
+$(BUILD_DIR)/src/os/initialize_isv.marker: OPTFLAGS := -O2
+$(BUILD_DIR)/src/log/delay.marker: MIPS_VERSION := -mips1 -o32 -KPIC
```
Comment (Thar0) [inline Makefile.ido:39]: Similarly, `-KPIC` doesn't really belong here

File: Makefile [hunk 18]
```diff
@@ -92,78 +93,60 @@ setup:
...
-
-MDEBUG_FILES := $(BUILD_DIR)/src/monutil.marker
-$(BUILD_DIR)/src/monutil.marker: CC := tools/ido/cc
-$(BUILD_DIR)/src/monutil.marker: ASFLAGS := -non_shared -mips2 -fullwarn -verbose -Xcpluscomm -G 0 -woff 516,649,838,712 -Wab,-r4300_mul -nostdinc -o32 -c
 
 $(BUILD_DIR)/%.marker: %.c
-	cd $(<D) && $(WORKING_DIR)/$(CC) $(CFLAGS) $(CPPFLAGS) $(OPTFLAGS) $(<F) $(INCLUDES) -o $(WORKING_DIR)/$(@:.marker=.o)
+	cd $(<D) && $(WORKING_DIR)/$(CC) $(CFLAGS) $(MIPS_VERSION) $(CPPFLAGS) $(OPTFLAGS) $(<F) $(IINC) -o $(WORKING_DIR)/$(@:.marker=.o)
+ifeq ($(COMPILER),ido)
+	mips-linux-gnu-objcopy --remove-section .mdebug $(@:.marker=.o)
```
Comment (Thar0) [inline Makefile:124]: The binutils prefix should be moved to a makefile variable e.g. `CROSS := mips-linux-gnu-` so it can be easily overriden

File: src/os/invaldcache.s [hunk 19]
```diff
@@ -3,23 +3,21 @@
...
 
 .text
-.set noreorder
 LEAF(osInvalDCache)
     blez a1, 3f
-    nop
     li t3, DCACHE_SIZE
     bgeu a1, t3, 4f
-     nop
+    # nop
```
Comment (Thar0) [inline src/os/invaldcache.s:10]: Personally I don't think comments for nop delay slots are really necessary, an empty line is equally good

**Review by Thar0 — COMMENTED** [review]
A couple of things

File: tools/remove_mdebug.sh [hunk 20]
```diff
@@ -0,0 +1,16 @@
...
+
+if [[ ! -d ".mdebug/" ]]
+then
+    mkdir .mdebug
+fi
+
+for f in $(find . -name '*.o' -not -path './.mdebug/*') ;
+do
+    cp $f .mdebug/$f
+    mips-linux-gnu-objcopy --remove-section .mdebug $f
```
[the remark this answers is not in the document]
Comment (hensldm) [reply tools/remove_mdebug.sh:13]: Hmm, I like your solution better, just stripping both with `$(COMPARE_OBJ)` rather than having a brand new script. The only modification I think would be saving the cmp object files to a hidden `.cmp` in the base folder, while creating the folder on `setup`. What are your thoughts?

File: Makefile.ido (hunk 16 above)
Comment (hensldm) [reply Makefile.ido:14]: Would you suggest a brand new variable to store `-non_shared` vs `-KPIC`? (if so do you have a good naming recommendation?)

Comment (Thar0) [reply Makefile.ido:14]: `PICFLAGS` maybe, although I don't think `CFLAGS += ...` is too bad either

File: src/voice/voicesetadconverter.c [hunk 21]
```diff
@@ -23,9 +24,7 @@ s32 __osVoiceSetADConverter(OSMesgQueue* mq, s32 channel, u8 data) {
             __osContLastCmd = CONT_CMD_SWRITE_VOICE;
             __osPfsLastChannel = channel;
 
-            for (i = 0; i < channel; i++) {
-                *ptr++ = 0;
-            }
+            for (i = 0; i < channel; i++) { *ptr++ = 0; }
```
[the remark this answers is not in the document]
Comment (hensldm) [reply src/voice/voicesetadconverter.c:27]: Possibly. I do want to do a follow PR to clean up some of the voice files that I've noticed. I'll look into it then. (From a quick glance there are a couple places outside of voice that I think the macro would want to be used, but probably won't work since it is not all on one line).

File: Makefile [hunk 22]
```diff
@@ -51,6 +49,12 @@ MARKER_FILES := $(O_FILES:.o=.marker)
 ifneq ($(NON_MATCHING),1)
 COMPARE_OBJ = cmp $(BASE_DIR)/$(@F:.marker=.o) $(@:.marker=.o) && echo "$(@:.marker=.o): OK"
 COMPARE_AR = cmp $(BASE_AR) $@ && echo "$@: OK"
+ifeq ($(COMPILER),ido)
+COMPARE_OBJ = $(CROSS)objcopy -p --strip-debug $(BASE_DIR)/$(@F:.marker=.o) $(BASE_DIR)/.cmp/$(@F:.marker=.cmp.o) && \
```
Comment (Thar0) [inline Makefile:53]: I didn't do this in the iQue branch but I thought the idea here was to make these files on setup at the same time as the `.cmp` dir?

File: Makefile [hunk 23]
```diff
@@ -92,78 +96,54 @@ setup:
...
-$(BUILD_DIR)/src/os/%.marker: ASFLAGS += -P
-$(BUILD_DIR)/src/gu/%.marker: ASFLAGS += -P
-$(BUILD_DIR)/src/libc/%.marker: ASFLAGS += -P
-$(BUILD_DIR)/src/rmon/%.marker: ASFLAGS += -P
+$(BUILD_DIR)/src/gu/parse_gbi.marker: GBIDEFINE :=
+$(BUILD_DIR)/src/gu/us2dex_emu.marker: GBIDEFINE :=
+$(BUILD_DIR)/src/gu/us2dex2_emu.marker: GBIDEFINE :=
+$(BUILD_DIR)/src/sp/sprite.marker: GBIDEFINE :=
+$(BUILD_DIR)/src/sp/spriteex.marker: GBIDEFINE :=
+$(BUILD_DIR)/src/sp/spriteex2.marker: GBIDEFINE :=
```
Comment (Thar0) [inline Makefile:120]: Can these be `F3D_GBI` or something instead of empty?

**Review by Thar0 — COMMENTED** [review]
Last things

File: Makefile (hunk 23 above)
Comment (hensldm) [reply Makefile:120]: They can, though I think the reason they are empty (even in current main) is they get defined in the c files directly. Would you still prefer having some kind of define?

**Review by Thar0 — APPROVED** [review]
🎉

**Review by AngheloAlf — APPROVED** [review]
